### DISCUSSION TITLE
DC Universe Online - Better Alternative Method for Extracting High-Resolution Textures?

### DISCUSSION THREAD

**Discussion by 0Curtain0** [gildor2/UEViewer#139] [Show and tell]
Hi, I want to apologize for posting this here. I tried to register on the forums ( https://www.gildor.org/smf/ ) but I I got the following error:
> An Error Has Occurred!
> Internal Forum Error

The forum has several topics about DC Universe Online, including the 2 latest topics from 2014: 
- https://www.gildor.org/smf/index.php/topic,2276.0.html 
- https://www.gildor.org/smf/index.php/topic,2412.0.html
<hr>Official statement from 2010:

> Implemented full support for the DC Universe Online.
> 
> Umodel has a few issues with this game:
>     some meshes will cause umodel to crash
>     TFC files are not working
> 
> Also, I have not tested animations (I have only 1 package with animations, it is working).
> 
> I have no plans to fix issues listed above until the game will be released (currently it is in a beta stage). Probably Sony will slightly change data formats (currently there are a lot of custom places in their engine).
<hr>UModel Changelog (2011):

> 17.03.2011
> - implemented DC Universe Online TFC texture support
> - implemented Homefront support (StaticMesh is not supported)
<hr>
I want to share my method for extracting the high-resolution textures, but it is difficult and tedious task. 
After I explain my method, I was hoping you can make the process easier in a future UModel update, if possible. 

The following page contains the models and textures I have extracted (except Cheetah): 
- https://www.deviantart.com/xcurtainx/gallery/65225525/dc-universe-online 

Since UModel isn't able to detect the location of the high resolution textures in the **TEXTUREFILECACHE** directory, I have to use a tool ([Raw TFC Exporter](https://forum.xentax.com/viewtopic.php?f=18&t=16033&sid=58b0dec8c1c835e4477d3062f0257096)) created by daemon1, to extract all the high-resolution textures from the TFC files. 
The problem is, there are so many textures files and they don't have unique filenames which makes it more difficult to identify the textures. 

Here is the method I've been using: 
I open UModel and locate the character I want to extract, **DCHRNPCICOVIL205.UPK** ([Lady Shiva](https://www.deviantart.com/xcurtainx/art/Lady-Shiva-DC-Universe-Online-PSK-723462714)). 

I open that **UPK** in UModel's 3D Viewer. 
I press **PGUP** and **PGDN** to see the associated textures, and that produces an error which is shown in the console log: 
```
Detected game root CHARACTER/NPC/ICONIC/VILLAIN (no recurse)
Found 196 game files (196 skipped) at path "CHARACTER/NPC/ICONIC/VILLAIN"
Scanned game directory in 0.0 sec, 602 allocs, 0.00 MBytes serialized in 0 calls.
Loading package: DCHRNPCICOVIL205.UPK Ver: 648/6405 Engine: 5859 Names: 234 Exports: 67 Imports: 19 Game: 800025
Full package loaded in 0.0 sec, 36 allocs, 0.00 MBytes serialized in 0 calls.
Loading SkeletalMesh3 CHRNPCICOVIL205_SKMESH from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_0 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_1 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_10 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_11 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_12 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_13 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_14 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_15 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_16 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_17 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_18 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_19 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_2 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_20 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_3 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_4 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_5 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_6 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_7 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_8 from package DCHRNPCICOVIL205.UPK
Loading SkeletalMeshSocket SkeletalMeshSocket_9 from package DCHRNPCICOVIL205.UPK
Loading MaterialInstanceConstant SHIVAHAIR from package DCHRNPCICOVIL205.UPK
Loading MaterialInstanceConstant LADYSHIVA from package DCHRNPCICOVIL205.UPK
Loading Texture2D LADY_SHIVA_DIFFUSE from package DCHRNPCICOVIL205.UPK
Loading Texture2D LADY_SHIVA_FX from package DCHRNPCICOVIL205.UPK
Loading Texture2D LADY_SHIVA_NORMAL from package DCHRNPCICOVIL205.UPK
Loading Material3 DNPCHAIREYE_MAT from package DCHRNPCICOVIL205.UPK
WARNING: Import(DMT_CHARACTERPRELOAD.FLATNORM) was not found
WARNING: Import(DMT_CHARACTERPRELOAD.WHITE) was not found
WARNING: Import(DMT_CHARACTERPRELOAD.FX_BASE) was not found
Loading Material3 DNPCDIFFUSE_MAT from package DCHRNPCICOVIL205.UPK
WARNING: Import(DMT_CHARACTERPRELOAD.BASICENV) was not found
Loaded in 0.0 sec, 62 allocs, 0.14 MBytes serialized in 3 calls.
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_FX: TFC file "CHARACTERSPECULAR" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_FX: TFC file "CHARACTERSPECULAR" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
Decompressing LADY_SHIVA_DIFFUSE: TFC file "Character" is missing
Decompressing LADY_SHIVA_NORMAL: TFC file "CHARACTERNORMALMAP" is missing
```
Near the end of the log, it shows the names of the textures and the missing TFC files they are located in: 
Decompressing **LADY_SHIVA_DIFFUSE**: TFC file "**Character**" is missing
Decompressing **LADY_SHIVA_NORMAL**: TFC file "**CHARACTERNORMALMAP**" is missing
Decompressing **LADY_SHIVA_FX**: TFC file "**CHARACTERSPECULAR**" is missing
<hr>

I then use **Raw TFC Exporter** to extract the **CHARACTER.TFC**, **CHARACTERNORMALMAP.TFC**, and the **CHARACTERSPECULAR.TFC**. 

The tool extracts all the textures as .DDS textures but with very different file names. 
Instead of LADY_SHIVA_NORMAL.DDS, you get something like this: CHARACTERNORMALMAP_123.DDS. 
Since the texture file names are not unique, it takes a long time to identify the correct textures for the character model. 
<hr>

If you know of a faster way to do this, please let me know. 
Once again, I apologize for posting this here. I'm still having difficulty registering on the forums.

**Comment by gildor2** [comment]
Thank you for sharing. Your method doesn't look suitable for any automated way, so it is not suitable for integration into UE Viewer even in theory. I hope the information still might be useful for other people. Added link to this issue to two threads mentioned above.

Regarding forum registration, I've integrated some spam protection system which is showing this message. If you're on dynamic IP, it is possible that your IP is in stopforumspam database (someone from your network got bad reputation there).

**Comment by n7sentinel** [comment]
UModel now is unable to export  DC Universe Online TFC texture correctly，do you still have Raw TFC Exporter ？

**Reply by HonorsoftEnterainment** [reply]
Here's a free TFC exporter package by Honorsoft

https://drive.google.com/file/d/14ygNR26cImdM1NyjUC8pMyViJBJqlKFh/view?usp=drivesdk

*Google Drive was flagging it as virus infected, but I used it many times without issue and have scanned all files with multiple software and it comes back clean.

GDrive seems to let you still download it though.

**Reply by PracticeBot13** [reply]
The link isn't available anymore. Please mirror.

**Reply by JesusFFM** [reply]
> The link isn't available anymore. Please mirror.

https://drive.google.com/file/d/16xEIaxLbU1DJq9-74s_kZwojQLzouitz/view?usp=sharing

Here you go, bud
